### ISSUE TITLE
Ordering of fonts in pubspec.yaml affects rendering on CanvasKit

### ISSUE THREAD

**Issue by timsneath** [flutter/flutter#62595]
## Steps to Reproduce

See https://github.com/rydmike/mr_font_test

When fonts are ordered:
```yaml
flutter:
  fonts:
    - family: Roboto
      fonts:
        - asset: assets/fonts/Roboto/Roboto-Thin.ttf
          weight: 100
        - asset: assets/fonts/Roboto/Roboto-Light.ttf
          weight: 300
        - asset: assets/fonts/Roboto/Roboto-Regular.ttf
          weight: 400
```
all `Text` widgets that don't have an explicit weight set are rendered with weight 100.

Reordering them so that the 400-weight font appears at the top of the list causes a line like:
```dart
const Text(
   'This page uses no specified font theme it should show up '
   'using the default font for each used platform.'),
```
to appear correctly.

This is a regression in CanvasKit compared to DomCanvas, where the font renders (as with mobile and desktop) with weight 400 in the above snippet regardless of the font ordering in `pubspec.yaml`.

(This bug is a tighter expression of #56319, now that we have pinpointed the issue.)

/cc @hterkelsen

**Comment by mdebbar** [comment]
I can only see thin text. I tried re-ordering the fonts, I tried setting `fontWeight` to `normal`. They all show thin text for me in CanvasKit.

```dart
Text(
  'This page uses no specified font theme it should show up '
  'using the default font for each used platform.',
  style: TextStyle(fontFamily: 'Roboto', fontWeight: FontWeight.normal),
)
```

I can confirm on the DOM backend, the text renders as expected.

**Comment by rydmike** [comment]
@mdebbar That sounds interesting. When I tried the re-ordering it worked fine for my case with the CanvasKit and CanvasKit+CanvasText font issues in https://github.com/flutter/flutter/issues/56319

Do you have a simple example shows it not working?

For me it worked fine when I in my pubspec.yaml file reordered my font assets, like so:

```yaml
flutter:
  fonts:
    - family: Roboto
      fonts:
        - asset: assets/fonts/Roboto/Roboto-Regular.ttf
          weight: 400
        - asset: assets/fonts/Roboto/Roboto-Thin.ttf
          weight: 100
        - asset: assets/fonts/Roboto/Roboto-Light.ttf
          weight: 300
```
For the actual case it was not just the Roboto font, a few other ones too and different styles of Roboto as well, like putting Roboto Italic Regular before the other italic weights.

Similar to what Tim did here for the issue demo code: https://github.com/rydmike/mr_font_test/pull/1/files but for my actual app. The last screenshot in issue https://github.com/flutter/flutter/issues/56319#issuecomment-666553493 also shows that it worked for the more complex app case and not just the issue demo.

**Comment by mdebbar** [comment]
@rydmike I'm testing on a brand new sample app:
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  MyApp();

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Center(
        child: Text(
          'hello world',
          style: TextStyle(fontFamily: 'Roboto', fontSize: 24.0),
        ),
      ),
    );
  }
}
```
I always see this, even if I re-order the fonts:
![image](https://user-images.githubusercontent.com/1278212/88982661-bead6100-d27d-11ea-9656-c84b657fba48.png)

I'll try this repo https://github.com/rydmike/mr_font_test.git to check if I see the same behavior.

**Comment by timsneath** [comment]
I think that's because you're using the system fonts? I can repro this with the same Dart code if I add the following to my pubspec.yaml:
```yaml
    - family: Roboto
      fonts:
        - asset: assets/fonts/Roboto/Roboto-Black.ttf
          weight: 900
        - asset: assets/fonts/Roboto/Roboto-Regular.ttf
          weight: 400
        - asset: assets/fonts/Roboto/Roboto-Light.ttf
          weight: 300
        - asset: assets/fonts/Roboto/Roboto-Thin.ttf
          weight: 100
        - asset: assets/fonts/Roboto/Roboto-Medium.ttf
          weight: 500
        - asset: assets/fonts/Roboto/Roboto-Bold.ttf
          weight: 700
        - asset: assets/fonts/Roboto/Roboto-ThinItalic.ttf
          weight: 100
          style: italic
        - asset: assets/fonts/Roboto/Roboto-RegularItalic.ttf
          weight: 400
          style: italic
        - asset: assets/fonts/Roboto/Roboto-LightItalic.ttf
          weight: 300
          style: italic
        - asset: assets/fonts/Roboto/Roboto-MediumItalic.ttf
          weight: 500
          style: italic
        - asset: assets/fonts/Roboto/Roboto-BoldItalic.ttf
          weight: 700
          style: italic
        - asset: assets/fonts/Roboto/Roboto-BlackItalic.ttf
          weight: 900
          style: italic
```

Here, the text is rendered differently depending on which weight is at the top of the list. For example:
![image](https://user-images.githubusercontent.com/2319867/88983229-4b0c5380-d27f-11ea-9531-a3ea933f8241.png)

**Comment by mdebbar** [comment]
I'm using `Roboto` too. Here's my `pubspec.yaml`:
```yaml
flutter:
  fonts:
    - family: Roboto
      fonts:
        - asset: assets/fonts/Roboto/Roboto-Thin.ttf
          weight: 100
        - asset: assets/fonts/Roboto/Roboto-Light.ttf
          weight: 300
        - asset: assets/fonts/Roboto/Roboto-Regular.ttf
          weight: 400

  uses-material-design: true
```
Maybe because `uses-material-design: true`? I'll try removing it.

**Comment by timsneath** [comment]
Weird -- I also have that line, just didn't include it. 

Here's my repo: https://github.com/timsneath/fonttest

as it stands, it shows Roboto Regular when I run with `flutter run -d chrome --release --dart-define=FLUTTER_WEB_USE_SKIA=true` on ToT master.

**Comment by mdebbar** [comment]
I was able to repro this in release mode. For some reason in debug mode it always renders thin text even when I re-order the fonts.

Anyway, I believe the problem is happening [here](https://github.com/flutter/engine/blob/c134e1658d566f45b2d65f813565cd182cbdd676/lib/web_ui/lib/src/engine/compositor/fonts.dart#L59-L75) and [here](https://github.com/flutter/engine/blob/c134e1658d566f45b2d65f813565cd182cbdd676/lib/web_ui/lib/src/engine/compositor/text.dart#L62-L64). Flutter refers to all of the Roboto-* fonts with the name `Roboto`. But our CanvasKit backend cares about the actual font names (e.g. `Robot-Thin`, `Roboto-Light`, etc). So we have a mapping from the flutter font name to the actual font name (`Roboto` -> `Roboto-Thin`). Since in this case `Roboto` refers to multiple actual fonts, we only pick the first one.

I tried to change the mapping so that a single font family maps to a list of actual font families (see code [here](https://github.com/mdebbar/engine/commit/ef21af510e4d2c67173433a8274a4384687cb594)). With that code, we are passing the entire list of font names to CanvasKit. But that didn't help. It seems like CanvasKit just picks the first font and doesn't care about the specified font weight. @hterkelsen any idea how this is supposed to work?
